src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,233 @@
+                            new AccessibleImageId(imageId.Guid, imageId.Id, "Temporary Automation Name"));
+                        filterCache[filter.DisplayText] = itemFilter;
+                        result.Add(itemFilter);
+                    }
+                }
+            }
+
+            return result.ToImmutableArray();
+        }
+
+        public async Task<object> GetDescriptionAsync(EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return string.Empty;
+            }
+
+            Workspace.TryGetWorkspace(triggerBuffer.AsTextContainer(), out var workspace);
+
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(triggerBuffer.AsTextContainer());
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+            var service = document.GetLanguageService<CompletionService>() as CompletionServiceWithProviders;
+            var description = await service.GetProvider(roslynItem).GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private CompletionServiceWithProviders GetCompletionService(ITextSnapshot snapshot)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var workspace = document.Project.Solution.Workspace;
+            return (CompletionServiceWithProviders)workspace.Services.GetLanguageServices(LanguageNames.CSharp).GetService<CompletionService>();
+        }
+
+        public Task HandleViewClosedAsync(ITextView view) => Task.CompletedTask;
+
+        public bool TryGetApplicableSpan(char typeChar, SnapshotPoint triggerLocation, out SnapshotSpan applicableSpan)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                applicableSpan = default;
+                return false;
+            }
+            
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                applicableSpan = default;
+                return false;
+            }
+
+            if (!document.TryGetText(out var sourceText))
+            {
+                applicableSpan = default;
+                return false;
+            }
+
+            // TODO: TypeChar of 0 means Invoke or InvokeAndCommitIfUnique. An API update will make this better.
```

Comment:
How would you like the API to look like?
I'd make the comment explicitly state that we have to provide a span in Invoke or InvokeAndCommitIfUnique scenarios